### DISCUSSION TITLE
Custom validation logic goes where?

### DISCUSSION THREAD

**Discussion by bluikko** [netbox-community/netbox#6865] [Q&A] [answered]
If I understand the custom validators right the purpose is to associate custom validators in the configuration file.

But if I want to write custom validation logic then where should that go?

**Comment by jeremystretch** [comment, marked as answer]
You can define custom validator classes in a separate file, but they must be imported into `configuration.py` so you can set `CUSTOM_VALIDATORS`: https://netbox.readthedocs.io/en/feature/customization/custom-validation/

**Reply by bluikko** [reply]
I read that and there was nothing about where custom validation logic should be put - only about associations. So if I understand right what you are saying: I could put those in the _configuration_ file (yeah it's python I get it but still), or add a line in the configuration file to import any other file.

I thought so but was hoping that I must be wrong about it.

I would definitely prefer a directory for these or even just a "magic" file name. But if I am alone thinking this then I'll just go with it.

Into which directory do other people put their custom validation logic files (I would have several)?

**Reply by jeremystretch** [reply]
> I thought so but was hoping that I must be wrong about it.

You were hoping that NetBox would give you _less_ flexibility to structure your code however you want?

**Reply by bluikko** [reply]
Looking at the `.gitignore` file - would putting them into `/netbox/scripts/` interfere with scripts?

**Reply by bluikko** [reply]
> > I thought so but was hoping that I must be wrong about it.
> 
> You were hoping that NetBox would give you _less_ flexibility to structure your code however you want?

Absolutely!! If there was a single separate file name for custom validation logic (not the associations), I might as well put the line to import other files there.

Edit: or just use that single file instead of the config file which would be difficult to maintain.

A directory would still provide maximal flexibility in my opinion and would follow reports and scripts.
